### PR TITLE
feat(forms): Allow a FormControl to use initial value as default.

### PR DESCRIPTION
**Pull request by dylhunn**
Allow a FormControl to be reset to its initial value. Provide this feature via a new option in a FormControlOptions interface, based on AbstractControlOptions.

Also, expose the default value as part of the public API. This is part of a feature that has been requested elsewhere (e.g. in #19747).

This was originally proposed as part of typed forms. As discussed in the GDE session (and after with akushnir/alxhub), it is likely better to just reuse the initial value rather than accepting an additional default.

It is desirable to land this separately in order to reduce the scope of the typed forms PR, and make it a types-only change.

Edit: This change will also help us in a long-term shift to reduce the use of `null` in Forms, and adopt safer defaults.

Pertains to issue #13721.
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
It is impossible to specify the value to which a control will reset ahead of time.

Issue Number: #13721

## What is the new behavior?
It is now possible to cause a control to reset to its initial value. This will be useful for simplifying the types in strongly typed reactive forms.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by dylhunn** [comment]
Presubmit is green: http://test/OCL:415607582:BASE:415607613:1639173976883:195d84c6

File: goldens/public-api/forms/forms.md [hunk 1]
```diff
@@ -361,6 +361,11 @@ export class FormControlName extends NgControl implements OnChanges, OnDestroy {
     static ɵfac: i0.ɵɵFactoryDeclaration<FormControlName, [{ optional: true; host: true; skipSelf: true; }, { optional: true; self: true; }, { optional: true; self: true; }, { optional: true; self: true; }, { optional: true; }]>;
 }
 
+// @public
+export interface FormControlOptions extends AbstractControlOptions {
+    resetToInitialValue?: boolean;
```
Comment (JoostK) [inline goldens/public-api/forms/forms.md:366]: Can we reasonably infer whether a control's type needs to include `|null` based on this option? Especially since it's `boolean`, the presence of the field is not sufficient to know that `null` has to be excluded.

Comment (dylhunn) [reply goldens/public-api/forms/forms.md:366]: Yes, we can! Here is a toy example of how it would work:

```
interface FormControlOptions {
  resetToDefault?: boolean;
}

declare interface FormControl<T> {
}

declare class FormControlImpl<T> {
  constructor(value: T, options?: FormControlOptions);
}

declare interface FormControlCtor {
  new<T>(
      value: T,
      options: FormControlOptions & {resetToDefault: true}): FormControl<T>;
  new<T>(
      value: T,
      options?: FormControlOptions): FormControl<T|null>;
}

export const FormControl: FormControlCtor = FormControlImpl as FormControlCtor;

// Usage

const fc = new FormControl(42, {resetToDefault: true});
```

Comment (dylhunn) [reply goldens/public-api/forms/forms.md:366]: (In the above example, if the compiler can prove that `resetToDefault` is always `true` then you get a `FormControl<T>`. In all other cases, you get `FormControl<T|null>`.)

Comment (JoostK) [reply goldens/public-api/forms/forms.md:366]: Thanks! Yeah, it depends on the literal true type being captured.

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

reviewed-for: public-api

**Comment by dylhunn** [comment]
@alxhub I renamed the key to `initialValueIsDefault` as you suggested on our call, as well as exporting the `defaultValue` field for public usage.

File: packages/forms/src/model.ts [hunk 2]
```diff
@@ -1236,6 +1252,9 @@ export class FormControl extends AbstractControl {
   /** @internal */
   _pendingChange: boolean = false;
 
+  /** @internal */
+  _resetToValue: any = null;
```
[the remark this answers is not in the document]
Comment (dylhunn) [reply packages/forms/src/model.ts]: Done, as well as exporting it.

File: packages/forms/src/model.ts [hunk 3]
```diff
@@ -1236,6 +1252,12 @@ export class FormControl extends AbstractControl {
   /** @internal */
   _pendingChange: boolean = false;
 
+  /**
+   * defaultValue contains the value that this `FormControl` will reset to.
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:1256]:
```suggestion
   * Contains the value that this `FormControl` will reset to.
   * See `initialValueIsDefault` config parameter for additional information.
```

File: packages/forms/src/model.ts [hunk 4]
```diff
@@ -146,6 +146,22 @@ export interface AbstractControlOptions {
...
+ *
+ * This interface extends all options from {@link AbstractControlOptions}, plus some options
+ * unique to `FormControl`.
+ *
+ * @publicApi
+ */
+export interface FormControlOptions extends AbstractControlOptions {
+  /**
+   * @description
+   * Whether to reset to the initial constructed value of the control.
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:160]: I think it'd be great to have an example or two (with and without a flag) on how this flag affects the behavior of the `reset` method.

File: packages/forms/src/model.ts [hunk 5]
```diff
@@ -1236,6 +1252,12 @@ export class FormControl extends AbstractControl {
   /** @internal */
   _pendingChange: boolean = false;
 
+  /**
+   * defaultValue contains the value that this `FormControl` will reset to.
+   * @publicApi
+   */
+  defaultValue: any = null;
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:1259]:
```suggestion
  readonly defaultValue: any = null;
```

File: packages/forms/test/form_control_spec.ts [hunk 6]
```diff
@@ -815,6 +815,17 @@ describe('FormControl', () => {
       expect(c.value).toBe(null);
     });
 
+    it('should reset to the initial value if specified in FormControlOptions', () => {
+      const c2 = new FormControl('foo', {initialValueIsDefault: true});
+      expect(c2.value).toBe('foo');
```
Comment (AndrewKushnir) [inline packages/forms/test/form_control_spec.ts:820]: Let's also add some assertions for the `defaultValue` field as well (since it's a part of the public API now).

**Review by AndrewKushnir — COMMENTED** [review]
@dylhunn thanks for additional changes 👍 I've left a few more comments.

Could you please also have a look at the commit message to reflect the recent changes (that `defaultValue` is a new public API field)?

File: packages/forms/src/model.ts (hunk 3 above)
Comment (dylhunn) [reply packages/forms/src/model.ts]: Done.

File: packages/forms/src/model.ts (hunk 5 above)
Comment (dylhunn) [reply packages/forms/src/model.ts]: Done.

File: packages/forms/test/form_control_spec.ts (hunk 6 above)
Comment (dylhunn) [reply packages/forms/test/form_control_spec.ts:820]: Done.

File: packages/forms/src/model.ts (hunk 4 above)
Comment (dylhunn) [reply packages/forms/src/model.ts:160]: Done.

File: packages/forms/src/model.ts [hunk 7]
```diff
@@ -146,6 +146,30 @@ export interface AbstractControlOptions {
...
+  /**
+   * @description
+   * Whether to reset to the initial constructed value of the control.
+   *
+   * ```ts
+   * const dog = new FormControl('spot');
+   * dog.reset(); // dog.value is null
+   *
+   * const cat = new FormControl('tabby', {initialValueIsDefault: true});
+   * cat.reset(); // cat.value is "tabby"
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:167]: Thanks for adding examples!
I'd propose adding one more where we pass a value to the `reset` function to illustrate that passed value has a precedence.

File: packages/forms/src/model.ts [hunk 8]
```diff
@@ -1236,6 +1260,12 @@ export class FormControl extends AbstractControl {
   /** @internal */
   _pendingChange: boolean = false;
 
+  /**
+   * Contains the value that this `FormControl` will reset to.
+   * @publicApi
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:1265]:
```suggestion
   * See `initialValueIsDefault` config parameter for additional information.
   * @publicApi
```

File: packages/forms/src/model.ts [hunk 9]
```diff
@@ -1340,8 +1372,10 @@ export class FormControl extends AbstractControl {
    * When false, no events are emitted.
    *
    */
-  override reset(formState: any = null, options: {onlySelf?: boolean, emitEvent?: boolean} = {}):
-      void {
+  override reset(formState: any = this.defaultValue, options: {
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:1375]: Let's also update the docs of this `reset` method to include a reference to the `initialValueIsDefault` config parameter.

**Review by AndrewKushnir — APPROVED** [review]
👍

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

File: packages/forms/src/model.ts (hunk 7 above)
Comment (dylhunn) [reply packages/forms/src/model.ts:167]: Done.

File: packages/forms/src/model.ts (hunk 8 above)
Comment (dylhunn) [reply packages/forms/src/model.ts:1265]: Done.

File: packages/forms/src/model.ts (hunk 9 above)
Comment (dylhunn) [reply packages/forms/src/model.ts:1375]: Done.

File: packages/forms/src/model.ts [hunk 10]
```diff
@@ -146,6 +146,36 @@ export interface AbstractControlOptions {
...
+   * const dog = new FormControl('spot');
+   * dog.reset(); // dog.value is null
+   *
+   * // If this flag is set, it will instead reset to the initial value.
+   * const cat = new FormControl('tabby', {initialValueIsDefault: true});
+   * cat.reset(); // cat.value is "tabby"
+   *
+   * // A value passed to reset always takes precedence.
+   * const fish = new FormControl('finn', {initialValueIsDefault: true});
+   * fish.reset('bubble'); // cat.value is "bubble"
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:173]:
```suggestion
   * fish.reset('bubble'); // fish.value is "bubble"
```

File: packages/forms/src/model.ts [hunk 11]
```diff
@@ -1323,8 +1362,9 @@ export class FormControl extends AbstractControl {
   }
 
   /**
-   * Resets the form control, marking it `pristine` and `untouched`, and setting
-   * the value to null.
+   * Resets the form control, marking it `pristine` and `untouched`, and resetting
+   * the value. The new will be the provided value (if passed), `null`, or the initial value if
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:1366]: Looks like a typo in the sentence?
```suggestion
   * the value. The new value will be the provided one (if passed), `null`, or the initial value if
```

File: packages/forms/src/model.ts (hunk 10 above)
Comment (dylhunn) [reply packages/forms/src/model.ts:173]: Fixed.

File: packages/forms/src/model.ts (hunk 11 above)
Comment (dylhunn) [reply packages/forms/src/model.ts:1366]: Fixed.

**Comment by dylhunn** [comment]
merge-assistance: test_aio appears to be erroneously failing:

```
Api pages should not show a "Properties" section if there are only internal properties
```

However, the page does have public properties, namely the new `defaultValue` property.

**Comment by dylhunn** [comment]
merge-assistance: when this PR is merged, it must be merged after #44479 for the tests to be green.

File: packages/forms/src/model.ts [hunk 12]
```diff
@@ -146,6 +146,36 @@ export interface AbstractControlOptions {
...
+ *
+ * This interface extends all options from {@link AbstractControlOptions}, plus some options
+ * unique to `FormControl`.
+ *
+ * @publicApi
+ */
+export interface FormControlOptions extends AbstractControlOptions {
+  /**
+   * @description
+   * Whether to reset to the initial constructed value of the control.
```
Comment (alxhub) [inline packages/forms/src/model.ts:160]: The docs here are accurate, but they describe the implementation behavior more than the concept. I took a stab at reframing them to talk more about "default":

> Whether to use the initial value used to construct the `FormControl` as its default value as well. If this option is `false` or not provided, the default value of a `FormControl` is `null`.
>
> When a `FormControl` is reset without an explicit value, its value reverts to its default value.

The example you have here of reset to default value vs explicit value is great, but I think it should go on the docs of the `reset()` method, not the docs of the constructor.

File: packages/forms/src/model.ts [hunk 13]
```diff
@@ -1228,8 +1258,17 @@ export abstract class AbstractControl {
  * @publicApi
  */
 export class FormControl extends AbstractControl {
+  /**
+   * Contains the value that this `FormControl` will reset to.
```
Comment (alxhub) [inline packages/forms/src/model.ts:1262]: I attempted to rephrase this too, to focus on the concept instead of the implementation.

> The default value of this `FormControl`, used whenever the control is reset without an explicit value.
>
> See {@link FormControlOptions#initialValueIsDefault} for more information on configuring a default value.

File: packages/forms/src/model.ts [hunk 14]
```diff
@@ -1265,6 +1303,9 @@ export class FormControl extends AbstractControl {
       // to `true` to allow that during the control creation process.
       emitEvent: !!this.asyncValidator
     });
+    if (isOptionsObj(validatorOrOpts) && validatorOrOpts.initialValueIsDefault) {
+      this.defaultValue = formState;
```
Comment (alxhub) [inline packages/forms/src/model.ts:1307]: Can't `formState` be an object here with a nested value? This seems wrong - default values shouldn't be `FormState`s I would think.

File: packages/forms/src/model.ts [hunk 15]
```diff
@@ -1323,8 +1364,9 @@ export class FormControl extends AbstractControl {
   }
 
   /**
-   * Resets the form control, marking it `pristine` and `untouched`, and setting
-   * the value to null.
+   * Resets the form control, marking it `pristine` and `untouched`, and resetting
+   * the value. The new value will be the provided value (if passed), `null`, or the initial value
+   * if `initialValueIsDefault` was set in the constructor via {@link FormControlOptions}.
```
Comment (alxhub) [inline packages/forms/src/model.ts:1369]: These docs would be a great place for the example above.

**Review by alxhub — DISMISSED** [review]

File: packages/forms/src/model.ts (hunk 12 above)
Comment (dylhunn) [reply packages/forms/src/model.ts]: I went ahead and adopted this change, and moved the examples.

File: packages/forms/src/model.ts (hunk 14 above)
Comment (dylhunn) [reply packages/forms/src/model.ts]: Nice catch! I have added a check for `formState`, as well as a test to verify it's handled properly.

File: packages/forms/src/model.ts (hunk 15 above)
Comment (dylhunn) [reply packages/forms/src/model.ts:1361]: I have moved the example!

File: packages/forms/src/model.ts (hunk 13 above)
Comment (dylhunn) [reply packages/forms/src/model.ts]: I have merged this rephrasing.

File: packages/forms/src/model.ts [hunk 16]
```diff
@@ -146,6 +146,24 @@ export interface AbstractControlOptions {
...
+ * unique to `FormControl`.
+ *
+ * @publicApi
+ */
+export interface FormControlOptions extends AbstractControlOptions {
+  /**
+   * @description
+   * Whether to use the initial value used to construct the FormControl as its default value as
+   * well. If this option is false or not provided, the default value of a FormControl is `null`.
+   * When a FormControl is reset without an explicit value, its value reverts to its default value.
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:162]: May be we can add a link to the `reset` function (to get more info and see examples)?

File: packages/forms/src/model.ts [hunk 17]
```diff
@@ -1265,6 +1291,13 @@ export class FormControl extends AbstractControl {
       // to `true` to allow that during the control creation process.
       emitEvent: !!this.asyncValidator
     });
+    if (isOptionsObj(validatorOrOpts) && validatorOrOpts.initialValueIsDefault) {
+      if (this._isBoxedValue(formState)) {
+        this.defaultValue = formState.value;
+      } else {
+        this.defaultValue = formState;
+      }
```
Comment (AndrewKushnir) [inline packages/forms/src/model.ts:1299]: Great catch 👍 I didn't mention this since there is a handling for the "boxed value" in the `reset` function (within `_applyFormState` code), but it would also update the disabled status, which is something we need to avoid (so special handling makes sense here).

File: packages/forms/test/form_control_spec.ts [hunk 18]
```diff
@@ -815,6 +815,46 @@ describe('FormControl', () => {
...
+      expect(c3.value).toBe('bar');
+      expect(c3.defaultValue).toBe(null);
+
+      c3.reset();
+      expect(c3.value).toBe(null);
+      expect(c3.defaultValue).toBe(null);
+    });
+
+    it('should look inside FormState objects for a default value', () => {
+      const c2 = new FormControl({value: 'foo', disabled: false}, {initialValueIsDefault: true});
```
Comment (AndrewKushnir) [inline packages/forms/test/form_control_spec.ts:845]: Let's also add a quick test with a "boxed value" and `disabled: true` to make sure that we do not reset the disabled state as well:

```
const c3 = new FormControl({value: 'foo', disabled: true}, {initialValueIsDefault: true});

expect(c3.status).toBe('DISABLED');

c3.enable();
c3.reset();

expect(c3.value).toBe('foo');
expect(c3.defaultValue).toBe('foo');
expect(c3.status).toBe('VALID'); // enabled
```

**Review by AndrewKushnir — APPROVED** [review]

File: packages/forms/test/form_control_spec.ts (hunk 18 above)
Comment (dylhunn) [reply packages/forms/test/form_control_spec.ts:845]: Done.

File: packages/forms/src/model.ts (hunk 16 above)
Comment (dylhunn) [reply packages/forms/src/model.ts]: Done.

**Comment by dylhunn** [comment]
(Rebased and all comments addressed)

**Comment by dylhunn** [comment]
@alxhub I just did one last push with all the comments addressed!

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 72092ebd26d6670a7dd02f6db5515d2afd27c4d6.
